### DISCUSSION TITLE
Regrouper les paquets publicodes et publicodes-react

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#192] [Ideas]
Dans https://github.com/betagouv/mon-entreprise/pull/1281 nous séparions publicodes en 2 paquets distincts : publicodes et publicodes-react. Le besoin est qu'une application qui n'utilise pas React, comme mesaidesvélo (ou un hypothétique site de l'Urssaf en Angular), puisse intégrer publicodes sans importer tout le code lié à React.

Au lieu de paquets séparés, je suis maintenant d'avis qu'une meilleure approche serait d'avoir un paquet unique mais avec 2 points d'entrée. 

Je ne suis pas 100% sûr, mais il me semble que ça peut être implémenté en définissant 2 `exports` dans le `package.json` :

```json
{
	"exports": {
		".": {
			"require": "./dist/core/index.cjs",
			"import": "./dist/core/index.js",
			"types": "./dist/core/index.d.ts"
		},
		"react": {
			"require": "./dist/react/index.cjs",
			"import": "./dist/react/index.js",
			"types": "./dist/react/index.d.ts"
		}
	},
}
```

Côté utilisateur on importerait les composants depuis `publicodes/react` (au lieu de du paquet `publicodes-react`) :

```js
import Publicodes from 'publicodes'
import { Documentation } from 'publicodes/react'
```

Je pense qu'un paquet avec 2 points d'entrée est un peu plus simple que 2 paquets distincts :

- les versions de `publicodes` et `publicodes-react` sont de toute façon toujours synchronisées, donc plus simple pour l'utilisateur de faire un `npm update publicodes` que de mettre à jour les 2 paquets
- process de build un peu plus simple de notre côté (on a par exemple eu le cas récemment où seul le paquet `publicodes` était publié sur NPM mais pas `publicodes-react`)
- c'est un détail mais dans le dépôt c'est un peu bizarre d'avoir des “workspaces” `core` et `react-ui` qui ne suivent pas le nom des paquets correspondants
- `yarn link` un peu plus simple depuis `mon-entreprise` s'il n'y a qu'un seul paquet à linker (et un seul build à lancer côté publicodes)
- plus facile à documenter : un seul README npm et non pas deux, moins d'explications sur la distinction des 2 paquets dans la documentation

Le seul inconvénient que je vois est que potentiellement le `node_modules` contiendra un peu plus de choses que nécessaire (`react`, `styled-components`) pour un utilisateur qui n'utiliserait pas `publicodes/react`. Mais ça ne concerne pas les applications bundlés donc ça ne semble pas trop critique. Et on peut potentiellement rendre ces dépendances optionnelles si c'est souhaitable ?

Par ailleurs dans un deuxième temps si on ne publie qu'un paquet unifié, il redevient possible de co-localiser la définition des mécanismes avec le composant React correspondant par exemple : 

```tsx
//mécansimes/somme.tsx

export default createMechanism({
  evaluate: (a, b) => a + b
})

export const jsx = function SommeExplanation({ explanation }) {
 return ...
}
```

Avec l'avantage que ce code qui est assez lié (typage des “explanations” notamment) se retrouve dans le même fichier. Normalement avec du code splitting, si on importe les fonctions jsx seulement depuis le module `publicodes/react` elles seront retirés du module `publicodes`.

**Comment by johangirod** [comment]
L'inconvénient principal selon moi, c'est la présence de `react` dans les `peerDependancies`. C'est typiquement le genre de chose que peut avancer une équipe un peu terre à terre pour ne pas utiliser publicodes (ce fut un des arguments avancés par l'Urssaf). 

Par ailleurs, si l'on ajoute d'autres visualisations par la suite (svelte, webcomponent ou anglular :trollface: ), cela nous obligera à ajouter d'autres "peerDependancies".

La solution actuelle est un chouilla galère, mais parait mieux fonctionner depuis les mises à jour du build. S'il y a encore des freins je suggère de les régler via du tooling un peu plus approprié (peut-être une commande pour mettre à jour tous les numéros de version en même temps ?).

Pour les noms des workspaces qui ne suivent pas ceux des paquets, je pense comme toi : c'est assez choquant. Je serai d'avis de les renommer.

**Reply by johangirod** [reply]
Après, pour nuancer, j'aime beaucoup l'idée d'avoir la visualisation à côté de la définition du mécanisme...
C'est pas évident évident...

**Reply by mquandalle** [reply]
Peut-être qu'on peut combiner les 2 approches :

- `export function evaluate` et `export function Explanation` dans le même fichier de définition d'un mécanisme
- paquet `publicodes` avec 2 points d'entrées : `publicodes` qui contient uniquement l'évaluation et `publicodes/jsx` qui ré-exporte les composants `Explanations` de chaque mécanisme. Le tree shaking permet de supprimer les composants JSX depuis le point d'entrée `publicodes` normal
- paquet `publicodes-react` qui importe le JSX des mécanismes depuis `publicodes/jsx` et qui inclut la dépendance à `react` et `react-dom` dans son `package.json` et exporte le composant `<Documentation />` comme aujourd'hui

Ainsi on peut définir un mécanisme dans un fichier unique (signature, évaluation, représentation), mais le paquet `publicodes` n'a pas `react` en dépendance et le point d'entrée du paquet bundlé ne contient pas le code des composants JSX grâce au tree shaking.

Si souhaité il devient même possible de mettre à jour seulement un des deux paquets : maj d'un mécanisme => maj de `publicodes`, maj de tout le reste du code relatif à l'affichage de la documentation (voire d'autres éléments d'UI comme les questions) => maj de `publicodes-react`. On peut aussi continuer à monter les 2 versions en même temps mais au moins, si un utilisateur fait uniquement un `npm update publicodes` et se retrouve avec des versions différentes de `publicodes` et `publicodes-react` la documentation ne cassera plus même si on a introduit un nouveau mécanisme ou modifié l'AST d'un mécanisme existant.

**Comment by mquandalle** [comment]
Voir aussi https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta qui permet de définir une “peer dependency” en `optional`
